From 10ad330317d765ba578cebacf2afc3616f0e1bd6 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Mon, 15 Jan 2024 08:45:01 +0100
Subject: [PATCH 0824/1141] media: dcmipp: pixelcap: call mutex_destroy on
 release and error handling

Add a mutex_destroy during the error handling within the function
dcmipp_pixelcap_ent_init and during the dcmipp_pixelcap_ent_release.

Change-Id: Ia1b207b6c48bd512daa2d4ba215c3b164d7aafa0
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/351139
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 .../platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c      | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
@@ -918,6 +918,7 @@ void dcmipp_pixelcap_ent_release(struct
 	struct dcmipp_pixelcap_device *vcap =
 		container_of(ved, struct dcmipp_pixelcap_device, ved);
 
+	mutex_destroy(&vcap->lock);
 	media_entity_cleanup(ved->ent);
 	vb2_video_unregister_device(&vcap->vdev);
 }
@@ -1123,7 +1124,7 @@ dcmipp_pixelcap_ent_init(struct device *
 	if (ret) {
 		dev_err(dev, "%s: vb2 queue init failed (err=%d)\n",
 			entity_name, ret);
-		goto err_clean_m_ent;
+		goto err_mutex_destroy;
 	}
 
 	/* Initialize buffer list and its lock */
@@ -1163,12 +1164,13 @@ dcmipp_pixelcap_ent_init(struct device *
 	if (ret) {
 		dev_err(dev, "%s: video register failed (err=%d)\n",
 			vcap->vdev.name, ret);
-		goto err_clean_m_ent;
+		goto err_mutex_destroy;
 	}
 
 	return &vcap->ved;
 
-err_clean_m_ent:
+err_mutex_destroy:
+	mutex_destroy(&vcap->lock);
 	media_entity_cleanup(&vcap->vdev.entity);
 err_clean_pads:
 	dcmipp_pads_cleanup(vcap->ved.pads);
